File: java/LongestPalindromicSubstring.java
/**
 * https://leetcode.com/problems/longest-palindromic-substring/description/
 */

public class LongestPalindromicSubstring {
    static String longestPalindrome(String s) {
        int len = s.length();
        int start=0, maxLength = 1;
        int l,h;

        for(int i=1; i<len; i++) {
            l = i-1;
            h = i;
            while (l >= 0 && h < len && s.charAt(l) == s.charAt(h)) {
                if (maxLength < (h-l+1)) {
                    start = l;
                    maxLength = h-l+1;
                }
                l--;
                h++;
            }

            l = i-1;
            h = i+1;
            while (l >= 0 && h < len && s.charAt(l) == s.charAt(h)) {
                if (maxLength < (h-l+1)) {
                    start = l;
                    maxLength = h-l+1;
                }
                l--;
                h++;
            }
        }
        return s.substring(start, start+maxLength);
    }

    public static void main(String[] args) {
        System.out.println(longestPalindrome("bb"));
    }
}

File: python3/submissions/longest_palindromic_substring.py
# https://leetcode.com/problems/longest-palindromic-substring/

class Solution:
    def longestPalindrome(self, s: str) -> str:
        if not s or len(s) < 2:
            return s
        max_so_far = 1
        max_idx = 0
        dp = set()
        length = len(s)
        for i in range(length):
            dp.add((i, i))
        for i in range(length-1):
            if s[i] == s[i+1]:
                max_so_far = 2
                max_idx = i
                dp.add((i, i+1))

        for k in range(3, length+1):
            for left in range(0, length-k+1):
                right = left+k-1
                if s[left] == s[right] and (left+1, right-1) in dp:
                    dp.add((left, right))
                    max_so_far = k
                    max_idx = left
        return s[max_idx:max_idx+max_so_far]